lib/spack/spack/fetch_strategy.py

```diff
@@ -1007,9 +1017,22 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
                     git(*args)
 
         # Init submodules if the user asked for them.
-        if self.submodules:
+        submodules = self.submodules
+        if callable(submodules):
+            submodules = submodules(self.package)
+            if isinstance(submodules, six.string_types):
+                submodules = shlex.split(submodules)
+            elif isinstance(submodules, (list, tuple)):
+                submodules = list(submodules)
+        if submodules:
             with working_dir(dest):
-                args = ['submodule', 'update', '--init', '--recursive']
+                args = ['submodule', 'init']
+                if isinstance(submodules, list):
```

Comment:
I would avoid lot of conditionals based on types. What do you think if we can restructure these lines to construct the git commands first, and finally invoke them? A simplifies sketch of the idea is:
```python
git_commands = []
if callable(submodules):
    ...
    git_commands.append(['submodule', 'init', '--'] +submodules)
    git_commands.append(['submodule', 'update', 'recursive'])
elif submodules is True:
    git_commands.append(['submodule', 'update', '--init', '--recursive'])

if not git_commands:
    return

with working_dir(dest):
    for args in git_commands:
        if not spack.config.get('config:debug'):
            args.insert(1, '--quiet')
        git(*args)
```